### PR TITLE
Refactoring GatedAutoCloseable and moving RecoveryState.Timer

### PR DESCRIPTION
**Pull request by kartg**
Signed-off-by: Kartik Ganesh <gkart@amazon.com>

### Description
This PR is one of the first steps in merging the feature branch `feature/segment-replication` back into `main` by re-PRing key changes. It covers part of the code changes in #2926 . The breakdown of the plan to merge segment-replication to main is detailed in #2355 

The changes included in this PR are as follows:

1. `GatedAutoCloseable` currently wraps a subclass of `RefCounted`. Segment replication adds another subclass, but this also wraps `RefCounted`. Both subclasses have the same shutdown hook - `decRef`. Therefore, `GatedAutoCloseable` has been made less generic (and renamed to `AutoCloseableRefCounted`) to increase code convergence.
2. The `RecoveryState.Timer` inner class has been moved to a top-level class. This will eventually be reused across both replication use-cases - peer recovery and segment replication.
3. In the `RecoveryTargetTests` test class, the test methods for `Timer` classes have been separated into distinct unit tests for each subclass. Common test logic is shared via a private method. The use of `RandomBoolean` has been removed to avoid non-deterministic code paths in testing. While doing so, the usage of `assertThat` (deprecated) has been replaced in these code paths.
4. Finally, a minor change removes the `readRecoveryState` API from `RecoveryState`, replacing it by a direct invocation of the constructor.

Segment replication design proposal - #2229  

### Issues Resolved
N/A
 
### Check List
- [ ] New functionality includes testing.
  - [x] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/indices/replication/common/Timer.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+import org.opensearch.common.io.stream.Writeable;
+import org.opensearch.common.unit.TimeValue;
+
+import java.io.IOException;
+
+/**
+ * A serializable timer that is used to measure the time taken for
+ * file replication operations like recovery.
+ */
+public class Timer implements Writeable {
```
Comment (andrross) [inline server/src/main/java/org/opensearch/indices/replication/common/Timer.java:22]: "Timer" is a _super_ generic name, used in lots of places (e.g. java.util.Timer, org.opensearch.search.profile.Timer, etc). Is there a more specific name that makes sense now that it is no longer an inner class? Maybe ReplicationTimer?

**Review by andrross — COMMENTED** [review]
Automated reviews are unhappy. Maybe missing a trailing newline?

**Comment by kartg** [comment]
Failure due to a flaky assertion on a timer. I'm currently looking into how we can improve these tests.

```
org.opensearch.indices.recovery.RecoveryTargetTests > testIndexTimer FAILED
    java.lang.AssertionError: expected:<20> but was:<14>
        at __randomizedtesting.SeedInfo.seed([1AB462596BA6129E:29DE4F529849686D]:0)
        at org.junit.Assert.fail(Assert.java:89)
        at org.junit.Assert.failNotEquals(Assert.java:835)
        at org.junit.Assert.assertEquals(Assert.java:647)
        at org.junit.Assert.assertEquals(Assert.java:633)
        at org.opensearch.indices.recovery.RecoveryTargetTests.doTimerTest(RecoveryTargetTests.java:180)
        at org.opensearch.indices.recovery.RecoveryTargetTests.testIndexTimer(RecoveryTargetTests.java:146)
```

**Comment by kartg** [comment]
Flaky test failure https://github.com/opensearch-project/OpenSearch/issues/1703

Refiring gradle check

**Review by dblock — APPROVED** [review]
This looks straightforward, thank you. Will leave it for @andrross to do another pass.

**Review by andrross — APPROVED** [review]

**Comment by dblock** [comment]
I labelled to backport to 2.x (2.1), nothing breaking here right?

**Comment by kartg** [comment]
> I labelled to backport to 2.x (2.1), nothing breaking here right?

Correct, nothing breaking; safe to backport to 2.x

**Pull request merged**